### DISCUSSION TITLE
[Subgraphs] - Using multiple query fields with different name per subgraph

### DISCUSSION THREAD

**Discussion by 2snEM6** [hayes/pothos#1004] [Q&A] [answered]
Let's say I have two subgraphs:

- admin
- public

Each subgraph must have a `login` query field, but currently Pothos complains about it: `Duplicate field definition for field login in Query`

How can I do that?

Thanks!

**Comment by hayes** [comment]
The subgraph plugin is exclusively for creating subsets of your graph.  You would only define 1 login field and include it as part of both subgraphs.

The subgraph plugin pre-dates apollos use of the term subgraph for federation, and the meaning there is a bit different than how it was used in Pothos.

**Reply by 2snEM6** [reply]
Thank you for the quick reply. I think then in my case I have to create 2 different builder instances and then call toSchema on each of them.

The problem relies on all the GraphQL Object Types. They import the `builder` instance, since they can be shared among different schemas.

Maybe I could create a main builder instance, register queries and mutations for the first schema, then delete query and mutations fields, then register queries and mutations of the second schema?

**Reply by hayes** [reply]
that might work, you could also check out the [add-graphql](https://pothos-graphql.dev/docs/plugins/add-graphql) plugin, build one schema, then use the add-graphql plugin to add the built pieces of your original schema to a new schema

**Reply by 2snEM6** [reply]
I understand `add-graphql` plugin does not add Query and Mutation roots correct?

**Reply by hayes** [reply]
add-graphql can add any GraphQL type to pothos I think, but the way you would use it would be to add everything except the query and mutation types

**Reply by 2snEM6** [reply]
Thanks, that makes sense. 

I was also wondering how to delete Query Fields from an existing builder instance and I assume it's not possible, since the `pendingFields` map is not public. Only the `getFields` is and it always displays an empty map

**Reply by hayes** [reply]
Yeah, currently this isn't really possible.  The functions that define fields aren't actually run until the schema gets built.  This helps with circular references/imports, but does mean that you can't really remove the fields since they aren't actually defined until you start the build process

**Reply by 2snEM6** [reply, marked as answer]
So what I did is make sure the schema is built before reusing the `builder` again. Then, creating a function that does this:

```typescript
  removeQueriesAndMutations(): this {
    this.configStore.typeConfigs.forEach((typeConfig) => {
      if (typeConfig.kind === 'Query' || typeConfig.kind === 'Mutation') {
        this.configStore.typeConfigs.delete(typeConfig.name);
      }
    });

    this.queryType({
      subGraphs: [...SUBGRAPHS],
      defaultSubGraphsForFields: [],
    });

    this.mutationType({
      subGraphs: [...SUBGRAPHS],
      defaultSubGraphsForFields: [],
    });

    const queryFields = this.configStore.getFields('Query', 'Object');
    queryFields.forEach((field) => {
      queryFields.delete(field.name);
    });

    const mutationFields = this.configStore.getFields('Mutation', 'Object');
    mutationFields.forEach((field) => {
      mutationFields.delete(field.name);
    });

    this.asyncTasks = [];

    return this;
  }
 ```

**Reply by hayes** [reply]
That will probably work for now, but just a heads up that a lot of these internal APIs are changing a lot in 4.0 (no eta, but there is a v4 branch with most of the expected changes)

**Reply by 2snEM6** [reply]
I extend the `Builder` class and added the following function:

```typescript
  subgraph(name: string) {
    this.currentSubgraph = name;
    return this.removeQueriesAndMutations();
  }
 ```
 
 Which in turn can be used as such:
 
 ```typescript
 const builder = new Builder();
 
 const subgraphA = builder.subgraph("admin");
 
 const schemaA = subgraphA.toSchema();
 
 const subgraphB = builder.subgraph("public");
 
 const schemaB = subgraphB.toSchema();
 ```

**Reply by 2snEM6** [reply]
> That will probably work for now, but just a heads up that a lot of these internal APIs are changing a lot in 4.0 (no eta, but there is a v4 branch with most of the expected changes)

Cool! thanks for the heads up. Looking forward to v4!
